registry/api/v2/names.go

```diff
@@ -16,9 +16,26 @@ const (
 )
 
 // RepositoryNameComponentRegexp restricts registry path component names to
-// start with at least one letter or number, with following parts able to
-// be separated by one period, dash or underscore.
-var RepositoryNameComponentRegexp = regexp.MustCompile(`[a-z0-9]+(?:[._-][a-z0-9]+)*`)
+// the allow valid hostnames according to: https://www.ietf.org/rfc/rfc2396.txt
+// with the following differences:
+//  1) It DOES NOT allow for fully-qualified domain names, which include a
+//    trailing '.', e.g. "google.com."
+//  2) It DOES NOT restrict 'top-level' domain labels to just alpha characters.
+//  3) It DOES allow for underscores to appear in the same situations as hyphens.
```

Comment:
I still don't agree with this statement, I think it is worth exploring allowing `_` at the same place as `.` rather than `-`. Since before we treated `_` exactly how `.` is handled now, it makes sense to limit the behavioral change to `-` since that is the target to be relaxed.